keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+objects and then select pods for evacuation. The pods should be selected by node (`nodeSelector`)
+and the pods should be gradually evacuated according to the workload they are running.
+Controllers can then implement the migration/termination either by reacting to the Evacuation API
+or by reacting to the NodeMaintenance API if they need more details.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I want to have an ability to manually switch between the
+maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
+observe the node drain via the API and check on its progress. I also want to be able to discover
+workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another. This also applies to
+applications with larger number of replicas that prefer to surge (upscale) pods first rather than
+downscale.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- This KEP depends on [Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+A misconfigured .spec.nodeSelector could select all the nodes (or just all master nodes) in the
+cluster. This can cause the cluster to get into a degraded and unrecoverable state.
+
+An admission plugin ([NodeMaintenance Admission](#nodemaintenance-admission)) is introduced to
+issue a warning in this scenario.
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain`: as we can see in the [Motivation](#motivation) section, kubectl is heavily used
+either manually or as a library by other projects. It is safer to keep the old behavior of this
+command. However, we will deprecate it along with all the library functions. We can print a
+deprecation warning when this command is used, and promote the NodeMaintenance. Additionally, pods
+that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`
+annotations will block eviction requests.
```

Comment:
TBH, I think that this overlap between eviction and evacuation could be very confusing for users.
If someone defines a PDB for an application, IMO they are entitled to expect that the application is protected during drain.

Reply:
> I think all tools built on top of the eviction API won't be able to drain Nodes anymore as soon as a single Pod opts into evacuation.

Not true, as the current design of the [Evacuation API](https://github.com/kubernetes/enhancements/pull/4565) stands, the eviction API will still fully work as before and will be fully supported. 

But you can opt into using NodeMaintenace+Evacuation API for more graceful pod terminations + node drain. 
We could easily enable this for example in the cluster autoscaler to observe the benefit of this feature. And it would also use eviction for pods who do not have evacuators assigned.

Reply:
My understanding is that as soon as a Pod opts into evacuation, eviction won't work anymore for this Pod ("Additionally, pods
that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`
annotations will block eviction requests").

This means as soon as a single Pod in a cluster opts into evacuation, all tools that have been build on top of eviction today (Cluster API, kubectl drain, many others) won't be able to drain the Node with this Pod anymore.